text/src/autogluon/text/automm/models/categorical_transformer.py

```diff
@@ -0,0 +1,275 @@
+import torch
+from torch import nn
+from torch import Tensor
+from typing import Any, Callable, Dict, List, Optional, Tuple, Type, Union, cast
+from ..constants import CATEGORICAL, LABEL, LOGITS, FEATURES
+from .ft_transformer import _TokenInitialization, CLSToken, Transformer
+
+
+class CategoricalFeatureTokenizer(nn.Module):
+    """
+        Transforms categorical features to tokens (embeddings).
+    """
+
+    category_offsets: Tensor
+
+    def __init__(
+        self,
+        num_categories: List[int],
+        d_token: int,
+        bias: Optional[bool] = True,
+        initialization: Optional[str] = 'normal',
+    ) -> None:
+        """
+        Args:
+            prefix:
+                The model prefix.
+            num_categories: 
+                The number of distinct values for each feature. For example,
+                :code:`num_categories=[3, 4]` describes two features: the first one can
+                take values in the range :code:`[0, 1, 2]` and the second one can take
+                values in the range :code:`[0, 1, 2, 3]`.
+            d_token: 
+                The size of one token.
+            bias: 
+                If `True`, for each feature, a trainable vector is added to the
+                embedding regardless of feature value. The bias vectors are not shared
+                between features.
+            initialization: 
+                Initialization policy for parameters. Must be one of
+                :code:`['uniform', 'normal']`. Let :code:`s = d ** -0.5`. Then, the
+                corresponding distributions are :code:`Uniform(-s, s)` and :code:`Normal(0, s)`. In
+                the paper [gorishniy2021revisiting], the 'uniform' initialization was
+                used.
+
+        References:
+            * [gorishniy2021revisiting] Yury Gorishniy, Ivan Rubachev, Valentin Khrulkov, Artem Babenko, "Revisiting Deep Learning Models for Tabular Data", 2021
+        """
+        super().__init__()
+        assert num_categories, 'num_categories must be non-empty'
+        assert d_token > 0, 'd_token must be positive'
+        
+        self.num_categories = num_categories
+
+        initialization_ = _TokenInitialization.from_str(initialization)
+
+        category_offsets = torch.tensor([0] + num_categories[:-1]).cumsum(0)
+
+        self.register_buffer('category_offsets', category_offsets, persistent=False)
+        self.embeddings = nn.Embedding(sum(num_categories), d_token)
+        self.bias = nn.Parameter(Tensor(len(num_categories), d_token)) if bias else None
+
+        for parameter in [self.embeddings.weight, self.bias]:
+            if parameter is not None:
+                initialization_.apply(parameter, d_token)
+
+    @property
+    def n_tokens(self) -> int:
+        """The number of tokens."""
+        return len(self.category_offsets)
```

Comment:
Consider to revise with `len(self.num_categories)` after you have removed category_offsets.